docs/docs/frontend-protocol.md

```diff
@@ -498,6 +498,7 @@ interface Op {
   op: "copy" | "skip" | "invalidate" | "update" | "ins"
   n: number  // number of lines affected
   lines?: Line[]  // only present when op is "update" or "ins"
+  ln?: number // the logical number for this line; null if this line is a soft break
```

Comment:
What does `this line` refer to here?

Reply:
Right, yea. This is definitely unclear.

This is the logical line number of the first line, if this is a "copy" op. (In theory this would also be the case for "update", but update doesn't exist and with the upcoming annotations work I think update will be removed.)

Something to note: if the first line in this "copy" is itself a continuation/wrapped line, `ln` will still refer to that line's logical number; this means that you'll need to increment `ln` to get the correct value for the first line that has a line number.

Reply:
Generally I'm not a fan of the complexity of the update protocol and I wonder if it is actually necessary. Did anyone ever benchmark the update protocol versus simply rerendering all visible lines for every update? I wouldn't be surprised if it wouldn't make any measurable difference in practice.

Reply:
as in not keeping a cache, and sending lines as needed?

I agree that this is an annoying amount of complexity. I think a longer-term solution might be to implement this all in rust as a library that clients can include.

This is guaranteed present if the op is a `copy`.

I hadn't thought of the situation where copy includes invalid lines; in this case I suspect this will just work, but I'd have to dig a bit.

Reply:
> as in not keeping a cache, and sending lines as needed?

I mean something similar to [what xray does](https://github.com/atom/xray/blob/7df278e3730eb66ee6ee73f39ff52249b326ef01/xray_core/src/buffer_view.rs#L992-L1058).

Reply:
That certainly does have an appealing simplicity. I've tended to defer to @raphlinus  on these types of architectural decisions, but I do think we're starting to bump up against some of the difficulties of our current approach, mostly that it involves a lot of ornate and finicky state tracking that's hard to maintain.